velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -97,60 +346,134 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
   probeRow_ += probeCnt;
-  if (probeRow_ == inputSize) {
-    probeRow_ = 0;
+  if (probeRow_ < input_->size()) {
+    return false;
+  }
+  probeRow_ = 0;
+  prevProbeCnt_ = 0;
+  do {
     ++buildIndex_;
-    if (buildIndex_ == buildData_->size()) {
-      buildIndex_ = 0;
-      input_.reset();
+  } while (!buildIndexAtEnd() && !buildData_.value()[buildIndex_]->size());
+  return buildIndexAtEnd();
+}
+
+RowVectorPtr NestedLoopJoinProbe::doMatching(vector_size_t probeCnt) {
+  VELOX_CHECK_NOT_NULL(input_);
+  VELOX_CHECK(!buildIndexAtEnd());
+
+  if (joinCondition_ == nullptr) {
+    return getCrossProduct(
+        probeCnt, outputType_, identityProjections_, buildProjections_);
+  }
+
+  auto filterInput = getCrossProduct(
+      probeCnt,
+      filterInputType_,
+      filterProbeProjections_,
+      filterBuildProjections_);
+
+  if (filterInputRows_.size() != filterInput->size()) {
+    filterInputRows_.resizeFill(filterInput->size(), true);
+  }
+  VELOX_CHECK(filterInputRows_.isAllSelected());
+
+  std::vector<VectorPtr> filterResult;
+  EvalCtx evalCtx(
+      operatorCtx_->execCtx(), joinCondition_.get(), filterInput.get());
+  joinCondition_->eval(0, 1, true, filterInputRows_, evalCtx, filterResult);
+  DecodedVector decodedFilterResult;
+  decodedFilterResult.decode(*filterResult[0], filterInputRows_);
+
+  const vector_size_t resultSize = decodedFilterResult.size();
+  auto rawProbeOutMapping =
+      initializeRowNumberMapping(probeOutMapping_, resultSize, pool());
+  auto rawBuildOutMapping =
+      initializeRowNumberMapping(buildOutMapping_, resultSize, pool());
+  auto probeIndices = probeIndices_->asMutable<vector_size_t>();
+  auto buildIndices = buildIndices_->asMutable<vector_size_t>();
+  int32_t outputSize{0};
+  for (auto i = 0; i < resultSize; ++i) {
+    if (!decodedFilterResult.isNullAt(i) &&
+        decodedFilterResult.valueAt<bool>(i)) {
+      rawProbeOutMapping[outputSize] = probeIndices[i];
+      rawBuildOutMapping[outputSize] = buildIndices[i];
+      outputSize++;
+    }
+  }
+  if (needsProbeMismatch(joinType_)) {
+    for (auto i = 0; i < outputSize; ++i) {
+      probeMatched_.setValid(rawProbeOutMapping[i], true);
+    }
+    probeMatched_.updateBounds();
+  }
+  if (needsBuildMismatch(joinType_)) {
+    for (auto i = 0; i < outputSize; ++i) {
+      buildMatched_[buildIndex_].setValid(rawBuildOutMapping[i], true);
     }
   }
-  return output;
-}
 
-bool NestedLoopJoinProbe::isFinished() {
-  return buildSideEmpty_ || (noMoreInput_ && input_ == nullptr);
+  if (!outputSize) {
+    return nullptr;
+  }
+  auto output = BaseVector::create<RowVector>(outputType_, outputSize, pool());
+  projectChildren(
+      output, input_, identityProjections_, outputSize, probeOutMapping_);
+  projectChildren(
+      output,
+      buildData_.value()[buildIndex_],
+      buildProjections_,
+      outputSize,
+      buildOutMapping_);
+  return output;
 }
 
-void NestedLoopJoinProbe::close() {
```

Comment:
Why remove this close method? Could you add to clear all the major objects that might use a lot memory here? We might abort operator early. Thanks!